plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3/TrinoS3FileSystem.java

```diff
@@ -596,27 +602,118 @@ public boolean rename(Path src, Path dst)
     public boolean delete(Path path, boolean recursive)
             throws IOException
     {
-        try {
-            if (!directory(path)) {
-                return deleteObject(keyFromPath(path));
+        String key = keyFromPath(path);
+        if (recursive) {
+            boolean deleted = deletePrefix(path);
+            if (deleted) {
+                deleteObject(key + DIRECTORY_SUFFIX);
+                return true;
+            }
+            else {
+                return deleteObject(key);
```

Comment:
In this path, if `key` exists as an object with content type `application/x-directory` then we would need to call `deleteObject(key + DIRECTORY_SUFFIX)` as well. (will require making a `getObjectMetadata(path, bucket, key)` call).